### PR TITLE
[SPARK-17106][SQL] Simplify the SubqueryExpression interface

### PR DESCRIPTION
**Pull request by hvanhovell**
## What changes were proposed in this pull request?

The current subquery expression interface contains a little bit of technical debt in the form of a few different access paths to get and set the query contained by the expression. This is confusing to anyone who goes over this code.

This PR unifies these access paths.
## How was this patch tested?

(Existing tests)

### CODE REVIEW

**Comment by hvanhovell** [comment]
cc @davies (we discussed this in PR https://github.com/apache/spark/pull/14548)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala [hunk 1]
```diff
@@ -56,30 +44,29 @@ trait ExecSubqueryExpression extends SubqueryExpression {
...
   override def nullable: Boolean = true
-  override def toString: String = executedPlan.simpleString
-
-  def withExecutedPlan(plan: SubqueryExec): ExecSubqueryExpression = copy(executedPlan = plan)
+  override def toString: String = plan.simpleString
+  override def withNewPlan(query: SubqueryExec): ScalarSubquery = copy(plan = query)
 
   override def semanticEquals(other: Expression): Boolean = other match {
-    case s: ScalarSubquery => executedPlan.sameResult(executedPlan)
+    case s: ScalarSubquery => plan.sameResult(s.plan)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala]: This fixes a small bug in `sematicEquals`.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala]: Good catch

**Comment by davies** [comment]
LGTM

**Pull request closed without merging**